src/main/java/pl/edu/mimuw/BigInt.java

```diff
@@ -27,16 +46,168 @@ public BigInt neg() {
+          reszta = sum / 10;
+          tab[i] = dig;
+        } else if (p2 >= 0) {
+          int sum = this.digits[p2] + reszta;
+          int dig = sum % 10;
+          reszta = sum / 10;
+          tab[i] = dig;
+        } else {
+          tab[i] = reszta;
+          reszta = 0;
+        }
+        p1--;
+        p2--;
+      }
+    } else {
+      // jedna liczba dodatnia druga ujemna
+      int p1 = d1 - 1;
+      int p2 = d2 - 1;
+      int bit = 0;
+      // od dluzszej odejmuje krotsza
+      if (d1 >= d2) {
+        que = this.isPositive;
+        for (int i = 0; i <= d1; i++) {
+          if (p1 >= 0 && p2 >= 0) {
+            int dig;
+            if (this.digits[p1] - bit < other.digits[p2]) {
+              if (this.digits[p1] == 0) {
+                dig = 10 - bit - other.digits[p2];
+              } else {
+                dig = this.digits[p1] + 10 - bit - other.digits[p2];
+              }
+              bit = 1;
+            } else {
+              dig = this.digits[p1] - bit - other.digits[p2];
+              bit = 0;
+            }
+            tab[i] = dig;
+          } else if (p1 >= 0) {
+            tab[i] = this.digits[p1] - bit;
+            bit = 0;
+          } else {
+            tab[i] = 0;
+          }
+          p1--;
+          p2--;
+        }
+      } else {
+        que = other.isPositive;
+        for (int i = 0; i <= d2; i++) {
+          if (p1 >= 0 && p2 >= 0) {
+            int dig;
+            if (other.digits[p2] - bit < this.digits[p1]) {
+              if (other.digits[p2] == 0) {
+                dig = 10 - bit - this.digits[p1];
+              } else {
+                dig = other.digits[p2] + 10 - bit - this.digits[p1];
+              }
+              bit = 1;
+            } else {
+              dig = other.digits[p2] - bit - this.digits[p1];
+              bit = 0;
+            }
+            tab[i] = dig;
+          } else if (p2 >= 0) {
+            tab[i] = other.digits[p2] - bit;
+            bit = 0;
+          } else {
+            tab[i] = 0;
+          }
+          p1--;
+          p2--;
+        }
+      }
+    }
+
+    // petla do odwracania tablicy, zeby liczba byla w poprawnej kolejnosci
+    int i;
+    int d;
+    int[] resultArr;
+    if (tab[tab.length - 1] == 0) {
+      d = tab.length - 1;
+      i = tab.length - 2;
+    } else {
+      d = tab.length;
+      i = tab.length - 1;
+    }
+    resultArr = new int[d];
+    for (int j = 0; j < d; j++) {
+      resultArr[j] = tab[i];
+      i--;
+    }
+
+    return new BigInt(resultArr, que);
   }
 
   public BigInt times(BigInt other) {
-    throw new IllegalStateException("TODO task 3: implement multiplication");
+    int d1 = this.digits.length;
+    int d2 = other.digits.length;
+    int[] tab = new int[d1 + d2];
+    int[] helper = new int[d1 + d2];
+    for (int i = 0; i < d1 + d2; i++) tab[i] = 0;
+    boolean que;
+    que = (this.isPositive && other.isPositive) || (!this.isPositive && !other.isPositive);
```

Comment:
polecam `final var que = (this.isPositive && other.isPositive) || (!this.isPositive && !other.isPositive);` - nie ma potrzeby 2 linii tutaj ;)